File: lib/services/settings_service.dart
/// Settings service for storing API keys locally
import 'package:shared_preferences/shared_preferences.dart';

class SettingsService {
  static SharedPreferences? _prefs;

  static Future<void> init() async {
    _prefs = await SharedPreferences.getInstance();
  }

  static SharedPreferences get prefs {
    if (_prefs == null) {
      throw Exception('SettingsService not initialized. Call init() first.');
    }
    return _prefs!;
  }

  // API Keys
  static String get deepgramApiKey => prefs.getString('deepgram_api_key') ?? '';
  static set deepgramApiKey(String value) => prefs.setString('deepgram_api_key', value);

  static String get openaiApiKey => prefs.getString('openai_api_key') ?? '';
  static set openaiApiKey(String value) => prefs.setString('openai_api_key', value);

  // Settings
  static String get language => prefs.getString('language') ?? 'en';
  static set language(String value) => prefs.setString('language', value);

  static String get openaiModel => prefs.getString('openai_model') ?? 'gpt-4o-mini';
  static set openaiModel(String value) => prefs.setString('openai_model', value);
  
  // Transcription mode: 'cloud' (Deepgram), 'whisper', or 'sherpa'
  static String get transcriptionMode => prefs.getString('transcription_mode') ?? 'cloud';
  static set transcriptionMode(String value) => prefs.setString('transcription_mode', value);
  
  // Whisper model size: 'tiny' or 'base'
  static String get whisperModelSize => prefs.getString('whisper_model_size') ?? 'tiny';
  static set whisperModelSize(String value) => prefs.setString('whisper_model_size', value);
  
  static bool get useLocalTranscription => transcriptionMode == 'sherpa' || transcriptionMode == 'whisper';
  static bool get useSherpa => transcriptionMode == 'sherpa';
  static bool get useWhisper => transcriptionMode == 'whisper';
  static bool get useDeepgram => transcriptionMode == 'cloud';
  
  // Audio source: 'omi' (default) or 'phone_mic'
  static String get audioSource => prefs.getString('audio_source') ?? 'omi';
  static set audioSource(String value) => prefs.setString('audio_source', value);
  
  static bool get useOmiDevice => audioSource == 'omi';
  static bool get usePhoneMic => audioSource == 'phone_mic';

  
  // Saved device for auto-connect
  static String get savedDeviceId => prefs.getString('saved_device_id') ?? '';
  static set savedDeviceId(String value) => prefs.setString('saved_device_id', value);
  
  static String get savedDeviceName => prefs.getString('saved_device_name') ?? '';
  static set savedDeviceName(String value) => prefs.setString('saved_device_name', value);
  
  static void clearSavedDevice() {
    prefs.remove('saved_device_id');
    prefs.remove('saved_device_name');
  }

  // Helper
  static bool get hasApiKeys => (deepgramApiKey.isNotEmpty || useLocalTranscription) && openaiApiKey.isNotEmpty;
  static bool get hasOpenAIKey => openaiApiKey.isNotEmpty;
  static bool get hasDeepgramKey => deepgramApiKey.isNotEmpty;
  
  // Notification settings
  static bool get notifyBatteryLow => prefs.getBool('notify_battery_low') ?? true;
  static set notifyBatteryLow(bool value) => prefs.setBool('notify_battery_low', value);
  
  static bool get notifyBatteryCritical => prefs.getBool('notify_battery_critical') ?? true;
  static set notifyBatteryCritical(bool value) => prefs.setBool('notify_battery_critical', value);
  
  static bool get notifyTaskReminders => prefs.getBool('notify_task_reminders') ?? true;
  static set notifyTaskReminders(bool value) => prefs.setBool('notify_task_reminders', value);
  
  static bool get notifyProcessing => prefs.getBool('notify_processing') ?? true;
  static set notifyProcessing(bool value) => prefs.setBool('notify_processing', value);
  
  // iCloud backup
  static bool get icloudBackupEnabled => prefs.getBool('icloud_backup_enabled') ?? false;
  static set icloudBackupEnabled(bool value) => prefs.setBool('icloud_backup_enabled', value);
}

File: OmiLocal/Sources/Services/SettingsService.swift
import Foundation

class SettingsService {
    static let shared = SettingsService()
    
    private let defaults = UserDefaults.standard
    
    init() {}
    
    var deepgramApiKey: String {
        get { defaults.string(forKey: "deepgram_api_key") ?? "" }
        set { defaults.set(newValue, forKey: "deepgram_api_key") }
    }
    
    var openaiApiKey: String {
        get { defaults.string(forKey: "openai_api_key") ?? "" }
        set { defaults.set(newValue, forKey: "openai_api_key") }
    }
    
    var language: String {
        get { defaults.string(forKey: "language") ?? "en" }
        set { defaults.set(newValue, forKey: "language") }
    }
    
    var openaiModel: String {
        get { defaults.string(forKey: "openai_model") ?? "gpt-4.1-mini" }
        set { defaults.set(newValue, forKey: "openai_model") }
    }
    
    var transcriptionMode: String {
        get { defaults.string(forKey: "transcription_mode") ?? "cloud" }
        set { defaults.set(newValue, forKey: "transcription_mode") }
    }
    
    var whisperModelSize: String {
        get { defaults.string(forKey: "whisper_model_size") ?? "tiny" }
        set { defaults.set(newValue, forKey: "whisper_model_size") }
    }
    
    var audioSource: String {
        get { defaults.string(forKey: "audio_source") ?? "omi" }
        set { defaults.set(newValue, forKey: "audio_source") }
    }
    
    var savedDeviceId: String? {
        get { defaults.string(forKey: "saved_device_id") }
        set { defaults.set(newValue, forKey: "saved_device_id") }
    }
    
    var savedDeviceName: String? {
        get { defaults.string(forKey: "saved_device_name") }
        set { defaults.set(newValue, forKey: "saved_device_name") }
    }
    
    var hasApiKeys: Bool {
        (!deepgramApiKey.isEmpty || transcriptionMode == "sherpa" || transcriptionMode == "whisper") && !openaiApiKey.isEmpty
    }
    
    var hasOpenAIKey: Bool {
        !openaiApiKey.isEmpty
    }
    
    var hasDeepgramKey: Bool {
        !deepgramApiKey.isEmpty
    }
    
    var notifyBatteryLow: Bool {
        get { defaults.bool(forKey: "notify_battery_low") }
        set { defaults.set(newValue, forKey: "notify_battery_low") }
    }
    
    var notifyBatteryCritical: Bool {
        get { defaults.bool(forKey: "notify_battery_critical") }
        set { defaults.set(newValue, forKey: "notify_battery_critical") }
    }
    
    var notifyTaskReminders: Bool {
        get { defaults.bool(forKey: "notify_task_reminders") }
        set { defaults.set(newValue, forKey: "notify_task_reminders") }
    }
    
    var notifyProcessing: Bool {
        get { defaults.object(forKey: "notify_processing") as? Bool ?? true }
        set { defaults.set(newValue, forKey: "notify_processing") }
    }
    
    var deepgramMinutesUsed: Double {
        get { defaults.double(forKey: "deepgram_minutes_used") }
        set { defaults.set(newValue, forKey: "deepgram_minutes_used") }
    }
    
    var openaiInputTokens: Int {
        get { defaults.integer(forKey: "openai_input_tokens") }
        set { defaults.set(newValue, forKey: "openai_input_tokens") }
    }
    
    var openaiOutputTokens: Int {
        get { defaults.integer(forKey: "openai_output_tokens") }
        set { defaults.set(newValue, forKey: "openai_output_tokens") }
    }
    
    func addDeepgramUsage(_ minutes: Double) {
        deepgramMinutesUsed += minutes
    }
    
    func addOpenAIUsage(input: Int, output: Int) {
        openaiInputTokens += input
        openaiOutputTokens += output
    }
    
    func resetUsageStats() {
        deepgramMinutesUsed = 0
        openaiInputTokens = 0
        openaiOutputTokens = 0
    }
    
    func clearSavedDevice() {
        defaults.removeObject(forKey: "saved_device_id")
        defaults.removeObject(forKey: "saved_device_name")
    }
}
